Editable collection project link

srallen opened the pull request:
Gravity Spy needs to be able to programmatically create collections, so being able to edit the project and default_subject links are required.

astopy commented:
You'll also need to rebase from master once #135 is merged, because `collection.links.projects` won't work at the moment.

srallen commented:
Huh that's interesting since I had Scotty try out this branch and he said it worked. I'll take a look after #135 is merged.

astopy commented:
It might have appeared to work if he'd tried something like this:

```
collection.project = new_project
collection.default_subject = new_subject
```

Because `Collection` objects don't have attributes by those names and Python will happily let you set new attributes on objects by default (that's probably something we should find a way to fix at some point I guess).

srallen commented:
@astopy I've rebased and added some class methods to update the default subject and project links modeled after what I see in the `SubjectSet` class. 

As for what he was doing and I'm asking for the script to confirm, but I think he was instantiating the class and editing the `links` attributes:
```
collection = Collection()
collection.links.project = project_id
collection.links.default_subject = subject_id
collection.save()
```

astopy commented:
@srallen That looks good at first glance, will review/test properly later. Is this also not needed right away like #134, or is this need sooner?

srallen commented:
@astopy this one is needed sooner as Scotty has already begun developing the feature using this branch.

srallen commented:
I've rebased this branch again so it should be in sync.

astopy commented on panoptes_client/collection.py:
```diff
@@ -12,7 +13,13 @@ class Collection(PanoptesObject):
+                'project',
+                'default_subject',
+            ),
+        },
```
I would revert this change, since these attributes shouldn't be edited directly.

astopy commented on panoptes_client/collection.py:
```diff
@@ -86,3 +93,34 @@ def _build_subject_list(self, subjects):
 
         return _subjects
+
+    def add_default_subject(self, subject):
```
Could you add docstrings to the new methods? (They're automatically included in the documentation.)

astopy commented on panoptes_client/collection.py:
```diff
@@ -86,3 +93,34 @@ def _build_subject_list(self, subjects):
+            _project_id = str(project)
+
+        self.http_post(
+            '{}/links/project/{}'.format(self.id, _project_id)
```
As above, the project ID should be in the request body here.

astopy commented on panoptes_client/collection.py:
```diff
@@ -86,3 +93,34 @@ def _build_subject_list(self, subjects):
+            _subject_id = str(subject)
+
+        self.http_post(
+            '{}/links/default_subject/{}'.format(self.id, _subject_id)
```
This isn't quite right. The subject ID should be given in the request body:

```python
self.http_post(
    '{}/links/default_subject'.format(self.id),
    json={'default_subject': subject_id},
)
```

astopy commented:
Almost there! The `add_default_subject` method works perfectly, but now that It think about it maybe we should rename it to `set_default_subject`? To make it clear from the name there can only be one default subject.

I'm getting a not found error from the API when I try to link a project:

```python
>>> c.link_to_project(7)
DEBUG:panoptes_client:json={'project': '7'}
DEBUG:urllib3.connectionpool:https://www.zooniverse.org:443 "POST /api/collections/21676/links/project HTTP/1.1" 404 56
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/src/panoptes-python-client/panoptes_client/collection.py", line 144, in link_to_project
    json={'project': _project_id},
  File "/usr/src/panoptes-python-client/panoptes_client/panoptes.py", line 506, in http_post
    json
  File "/usr/src/panoptes-python-client/panoptes_client/panoptes.py", line 320, in post
    endpoint=endpoint
  File "/usr/src/panoptes-python-client/panoptes_client/panoptes.py", line 221, in json_request
    json_response['errors']
panoptes_client.panoptes.PanoptesAPIException: Not Found
```

At a guess maybe it should be POSTing to `{}/links/projects` (note the `s` on the end)? Hard to tell because this endpoint doesn't seem to be in the Panoptes docs.

srallen commented:
I've updated the name of the method. I also updated the project link route to be plural, but I'm not sure about that either. This line in Panoptes's routes file implies only subjects and default_subject would have a route:
https://github.com/zooniverse/Panoptes/blob/cc63b57f72b5ffc9152885eae85037d823f70323/config/routes.rb#L125

On PFE we just include it in the json payload for the POST for creating the collection. We don't have any UI for editing the project link directly.

astopy commented:
```python
>>> c = Collection.where(owner='astopy').next()
DEBUG:panoptes_client:params={'owner': 'astopy'}
DEBUG:urllib3.connectionpool:https://www.zooniverse.org:443 "GET /api/collections?owner=astopy HTTP/1.1" 200 1151
>>> c
<Collection 77687>
>>> c.link_to_project(7)
DEBUG:panoptes_client:json={'project': '7'}
DEBUG:urllib3.connectionpool:https://www.zooniverse.org:443 "POST /api/collections/77687/links/projects HTTP/1.1" 404 56
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/src/panoptes-python-client/panoptes_client/collection.py", line 144, in link_to_project
    json={'project': _project_id},
  File "/usr/src/panoptes-python-client/panoptes_client/panoptes.py", line 506, in http_post
    json
  File "/usr/src/panoptes-python-client/panoptes_client/panoptes.py", line 320, in post
    endpoint=endpoint
  File "/usr/src/panoptes-python-client/panoptes_client/panoptes.py", line 221, in json_request
    json_response['errors']
panoptes_client.panoptes.PanoptesAPIException: Not Found
>>> 
```

@camallen Is there actually an endpoint for linking collections to projects? If so, what is it?

srallen commented:
@camallen Scotty is working on being able to programmatically create collections linked to Gravity Spy based on an algorithm search for similar glitch types. So, he just needs to be able to link a collection to a project, Gravity Spy, when it is created.

astopy commented:
Hmm, so now I'm confused about the right way to do this. The Panoptes API docs for collections seem to be completely wrong/out of date when it comes to both default subjects and project links.

@srallen What does the JSON look like that PFE submits when it creates collections?

camallen commented:
The `default_subject` can be set via the links on update & create. The linked projects can only be set by the links on create. Linking a project the resource has been created will not work.

srallen commented:
@astopy here's what the payload on create looks like on PFE when in a project context:
```
{
  display_name: "Collection display name",
  links: {project: "335", subjects: ["4220"]},
  private: false
}
```

astopy commented:
OK, so I guess what we need to do is remove the `link_to_project` function and reinstate 9aced2951687521625926fd99c49e39fc0fa76f7 to make the project link editable on create. The default subject stuff can stay as-is though I think.

srallen commented:
@astopy I've added to projects link back in as an editable attribute and rebase. This is ready for review again.

astopy commented:
Cool, that all works. The API is a little inconsistent with project vs projects, but that's not really an issue with the client:

```python
>>> p = Project(7)
>>> c = Collection()
>>> c.links.project = p
>>> c.display_name = 'Created from Python'
>>> c.save()
>>> c.links.project
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/src/panoptes-python-client/panoptes_client/panoptes.py", line 782, in __getattr__
    linked_object = self.parent.raw['links'][name]
KeyError: 'project'
>>> c.links.projects
[<Project 7>]
```

camallen commented:
why not just ensure the `c.links.projects` assignment is a list with length 1?

astopy commented:
We could do something like that to work around it in the client, yeah, though the front end specifies a single project link too (see @srallen's comment above). Maybe that's the expected behaviour, it just surprised me because I hadn't seen it before.
